### PR TITLE
Do not implement unsafe auto traits for types with unsafe fields

### PR DESCRIPTION
**Pull request by jswrenn**
If a type has unsafe fields, its safety invariants are not simply the conjunction of its field types' safety invariants. Consequently, it's invalid to reason about the safety properties of these types in a purely structural manner — i.e., the manner in which `auto` traits are implemented. Consequently, auto implementations of unsafe auto traits should not be generated for types with unsafe fields.

Tracking: #132922

r? @compiler-errors

### CODE REVIEW

**Review by compiler-errors — CHANGES_REQUESTED** [review]
Please adjust this logic in the new trait solver too. Also, there's no test as far as I can tell.

**Comment by compiler-errors** [comment]
Also make sure the test exercises the behavior in both the old and new solvers. You can do something like:

```
//@ revisions: current next
//@[next] compile-flags: -Znext-solver
```

**Comment by jswrenn** [comment]
Whoops, somehow failed to push the test I wrote. It's here: https://github.com/rust-lang/rust/compare/e6e322f79713b156d43d77be6135463ec2d4b6e7..baec50ee0b8afc7433c3bd2f2b862a21addd2f57

File: tests/ui/unsafe-fields/auto-traits.rs [hunk 1]
```diff
@@ -0,0 +1,23 @@
+//@ compile-flags: --crate-type=lib
+
+#![feature(auto_traits)]
+#![feature(unsafe_fields)]
+#![allow(dead_code, incomplete_features, unconditional_recursion)]
```
Comment (compiler-errors) [inline tests/ui/unsafe-fields/auto-traits.rs:5]: Please remove the `unconditional_recursion`. No reason not to just split this out into a `fn main() {}` or something that just calls both of the functions.

Comment (compiler-errors) [reply tests/ui/unsafe-fields/auto-traits.rs:5]: No need for `dead_code`, that's enabled in the UI test suite unconditionally.

Comment (jswrenn) [reply tests/ui/unsafe-fields/auto-traits.rs]: [Done](https://github.com/rust-lang/rust/compare/baec50ee0b8afc7433c3bd2f2b862a21addd2f57..9d604ce5704f06017fa06a1ebacd496f9cebb4eb#diff-f20103ee4791c97758b52ea3c6d14d523b123dc658f33307bfe72217089d48f8L5-R7)

**Comment by jswrenn** [comment]
> Please adjust this logic in the new trait solver too.

[Done](https://github.com/rust-lang/rust/compare/baec50ee0b8afc7433c3bd2f2b862a21addd2f57..9d604ce5704f06017fa06a1ebacd496f9cebb4eb#diff-1121b327882c6099c683bcfce42ad82a9bde00fab9e30058353951741e256bde)

> Also make sure the test exercises the behavior in both the old and new solvers. You can do something like:
> 
> ```
> //@ revisions: current next
> //@[next] compile-flags: -Znext-solver
> ```

[Done](https://github.com/rust-lang/rust/compare/baec50ee0b8afc7433c3bd2f2b862a21addd2f57..9d604ce5704f06017fa06a1ebacd496f9cebb4eb#diff-f20103ee4791c97758b52ea3c6d14d523b123dc658f33307bfe72217089d48f8R2-R3)

File: compiler/rustc_type_ir/src/inherent.rs [hunk 2]
```diff
@@ -136,6 +136,9 @@ pub trait Ty<I: Interner<Ty = Self>>:
         matches!(self.kind(), ty::FnPtr(..))
     }
 
+    /// Checks whether this type directly contains unsafe fields.
```
Comment (compiler-errors) [inline compiler/rustc_type_ir/src/inherent.rs:139]:
```suggestion
    /// Checks whether this type is an ADT that has unsafe fields.
```

**Review by compiler-errors — APPROVED** [review]

**Pull request merged**